cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
 	return "switch_root [-h] [-V]\nswitch_root newroot init"
 }
 
-// littleDoctor recursively deletes everything at "path" that
-// is in the same file system as "fs".
-func littleDoctor(path string, fsType int64) error {
-	var pathFS syscall.Statfs_t
+// get device for file descriptor
+func getDev(fd int) (dev uint64, err error) {
+	var stat syscall.Stat_t
 
-	if err := syscall.Statfs(path, &pathFS); err != nil {
-		return err
+	if err := syscall.Fstat(fd, &stat); err != nil {
+		return 0, err
 	}
 
-	if int64(pathFS.Type) != fsType {
-		return nil
-	}
+	return stat.Dev, nil
+}
 
-	file, err := os.Open(path)
-	if err != nil {
-		return fmt.Errorf("Could not open %s: %v", path, err)
+// check for the symlink bit
+func isSymlink(mode uint32) bool {
+	// numbers come from `man fstatat`
+	return (mode & 0170000) == 0120000
```

Comment:
`mode & syscall.S_IFMT == syscall.S_IFLNK`

Reply:
Grrr I grepped for this, but only for the octal number, not hex. Shame on me!